Sidepanel submodule: Recreate tree only at structure changes

gerhardol opened the pull request:
Fixes #8871
Related to #8886

## Proposed changes

The submodules in the sidepanel are recreated every time git-status is updated.
Before #8867, that occurred every 5th second. A gif show the behavior there.
#8867 restores the behavior to 3.4 and refreshing every 15th second.
#8886 proposes increasing the time further to 30 seconds.

This PR only recreate the submodule structure when there are structure changes.

No screenshot - but not much to see.

## Test methodology <!-- How did you ensure quality? -->

Manual

## Test environment(s) <!-- Remove any that don't apply -->

- Git Extensions 33.33.33
- Build 10df80cb1fe2ff8c07a30be3525635161f04c42c
- Git 2.30.1.windows.1
- Microsoft Windows NT 10.0.19041.0
- .NET Framework 4.8.4300.0
- DPI 96dpi (no scaling)

----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

gerhardol commented on GitUI/BranchTreePanel/RepoObjectsTree.Nodes.Submodules.cs:
```diff
@@ -260,23 +263,67 @@ private void OnStatusUpdated(SubmoduleStatusEventArgs e)
+                        {
+                            if (infos.ContainsKey(node.Info.Path))
+                            {
+                                node.Info = infos[node.Info.Path];
```
An option here is to check if the status has changed and only update the required nodes.
(The tooltip requires an additional git command, so there is some benefit).
Will increase the complexity though.

mstv commented on GitUI/CommandsDialogs/FormBrowse.cs:
```diff
@@ -2818,78 +2822,105 @@ private void SubmoduleStatusProvider_StatusUpdated(object sender, SubmoduleStatu
+            return newItems;
+        }
+
+        private async Task UpdateSubmoduleMenuStatus(SubmoduleInfoResult result, CancellationToken cancelToken)
```
```suggestion
        private async Task UpdateSubmoduleMenuStatusAsync(SubmoduleInfoResult result, CancellationToken cancelToken)
```

gerhardol replied:
yes

mstv commented on GitUI/CommandsDialogs/FormBrowse.cs:
```diff
@@ -2818,78 +2822,105 @@ private void SubmoduleStatusProvider_StatusUpdated(object sender, SubmoduleStatu
+            infos[result.TopProject.Path] = result.TopProject;
+            foreach (var item in _currentSubmoduleMenuItems)
+            {
+                var path = (string)item.Tag;
```
Use `as` syntax?

gerhardol replied:
yes, the next line nullchecks which is what we are after

mstv replied:
I thought with `as`, it would become shorter and `var` would be saved.

mstv commented on GitUI/CommandsDialogs/FormBrowse.cs:
```diff
@@ -2818,78 +2822,105 @@ private void SubmoduleStatusProvider_StatusUpdated(object sender, SubmoduleStatu
+                else
+                {
+                    // structure no longer matching
+                    Debug.Assert(true, $"Status info with {1 + result.AllSubmodules.Count} records do not match current nodes ({_currentSubmoduleMenuItems.Count})");
```
Would it log a message?

gerhardol replied:
This should never happen but is not a catastrophe if it does, so I believe this is OK to not log

mstv replied:
`Debug.Fail(...)` or `Debug.Assert(false, ...)`?

gerhardol replied:
I changed to Debug.Fail

mstv reviewed (commented):
Looks sensible.

gerhardol commented:
@RussKie What about including this in 3.5?
The behavior in 3.5 is probably better with the latest changes but this seem to address most of the flickering seen in #8871

RussKie commented:
I think we can, I'll review in the next few days.

On Wed, 3 Mar. 2021, 8:55 am Gerhard Olsson, <notifications@github.com>
wrote:

> @RussKie <https://github.com/RussKie> What about including this in 3.5?
> The behavior in 3.5 is probably better with the latest changes but this
> seem to address most of the flickering seen in #8871
> <https://github.com/gitextensions/gitextensions/issues/8871>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/gitextensions/gitextensions/pull/8902#issuecomment-789247667>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABBTEXQ7DMLPUUOHQ6KITA3TBVNFFANCNFSM4YKNHARA>
> .
>

RussKie reviewed (commented):
It would be helpful if I'd sent my review... It looks like I never have.

Overall looks sensible, haven't run.

gerhardol commented on GitCommands/Submodules/SubmoduleStatusEventArgs.cs:
```diff
@@ -7,12 +7,15 @@ public class SubmoduleStatusEventArgs : EventArgs
     {
         public SubmoduleInfoResult Info { get; }
 
+        public bool Structure { get; }
```
xml is already added

ok name?
```suggestion
        public bool StructureUpdated { get; }
```

RussKie commented on GitCommands/Submodules/SubmoduleStatusEventArgs.cs:
```diff
@@ -7,12 +7,18 @@ public class SubmoduleStatusEventArgs : EventArgs
         public SubmoduleInfoResult Info { get; }
 
+        /// <summary>
+        /// First update, updating the submodule structure. Status my be updated later.
```
```suggestion
        /// First update of the submodule structure. Status of the submodule will be updated asynchronously.
```

RussKie reviewed (approved):
With a suggestion

mstv commented on GitUI/BranchTreePanel/RepoObjectsTree.Nodes.Submodules.cs:
```diff
@@ -260,23 +263,67 @@ private void OnStatusUpdated(SubmoduleStatusEventArgs e)
+
+                    if (_currentNodes is not null)
+                    {
+                        var token = cts?.Token is not null ? cts.Token : e.Token;
```
```suggestion
                        var token = cts?.Token ?? e.Token;
```

mstv commented on GitUI/CommandsDialogs/FormBrowse.cs:
```diff
@@ -2828,78 +2832,105 @@ private void SubmoduleStatusProvider_StatusUpdated(object sender, SubmoduleStatu
+            if (infos.Count > 0)
+            {
+                // structure no longer matching
+                Debug.Assert(true, $"Status info has {infos.Count} records remaining after matching current nodes ({_currentSubmoduleMenuItems.Count}/{result.AllSubmodules.Count})");
```
ditto

mstv commented:
One of the assertions is not fulfilled on startup - both with submodules hidden or shown in the left panel.

![errors](https://user-images.githubusercontent.com/36601201/110248892-9b653580-7f73-11eb-8fb7-7fe63e98ab6f.png)

gerhardol commented:
> One of the assertions is not fulfilled on startup - both with submodules hidden or shown in the left panel.

Fixed in 42e13806b453e87a9925713f943d22f33bfec889 - changed this assert in a similar way as the other you requested and missed this

gerhardol commented:
> Fixed in [42e1380](https://github.com/gitextensions/gitextensions/commit/42e13806b453e87a9925713f943d22f33bfec889) - changed this assert in a similar way as the other you requested and missed this

Unfortunately not for submodules...
Squashed existing, added a new fixup for this
Also setting the icon for current submodule (extra menu item, it not top mpdule)

mstv reviewed (approved):
I haven't noticed issues. Thank you.

gerhardol commented:
Only rebased to merge, no changes
